### PR TITLE
[autoscaler/tune] Optional YAML Fields + Fix Pretty Printing for Tune

### PR DESCRIPTION
**Pull request by richardliaw**
## What do these changes do?

Simplifies YAML config to reduce user burden. 

TODOS: 
 - [x] Add docs
 - [x] Add tests

## Addresses

#1533, #1540

### CODE REVIEW

File: python/ray/tune/logger.py [hunk 1]
```diff
@@ -16,6 +16,11 @@
     tf = None
     print("Couldn't import TensorFlow - this disables TensorBoard logging.")
 
+try:
```
Comment (richardliaw) [inline python/ray/tune/logger.py]: this could go in a different PR

File: python/ray/autoscaler/aws/default.yaml [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+#    "/path1/on/remote/machine": "/path1/on/local/machine",
+#    "/path2/on/remote/machine": "/path2/on/local/machine",
+}
+
+# List of shell commands to run to set up nodes.
+setup_commands:
+    # Note: if you're developing Ray, you probably want to create an AMI that
+    # has your Ray repo pre-cloned. Then, you can replace the pip installs
+    # below with a git checkout <your_sha> (and possibly a recompile).
+    - pip install -U ray==0.3.1
```
Comment (richardliaw) [inline python/ray/autoscaler/aws/default.yaml]: This actually probably doesn't work; hopefully will be fixed in #1606 and then we can update this.

File: doc/source/autoscaling.rst [hunk 3]
```diff
@@ -66,10 +66,14 @@ The Ray autoscaler also reports per-node status in the form of instance tags. In
...
+You are encouraged to copy the example YAML file and modify it to your needs. This may include adding additional setup commands to install libraries or sync local data files.
+
+.. note:: After you have customized the nodes, it is also a good idea to create a new machine image (AMI) and use that in the config file. This reduces worker setup time, improving the efficiency of auto-scaling.
 
 The setup commands you use should ideally be *idempotent*, that is, can be run more than once. This allows Ray to update nodes after they have been created. You can usually make commands idempotent with small modifications, e.g. ``git clone foo`` can be rewritten as ``test -e foo || git clone foo`` which checks if the repo is already cloned first.
 
+Most of the example YAML file is optional. Here is a `reference minimal YAML file <https://github.com/ray-project/ray/tree/master/python/ray/autoscaler/aws/minimal.yaml>`__, and you can find the defaults for `optional fields in this YAML file <https://github.com/ray-project/ray/tree/master/python/ray/autoscaler/aws/default.yaml>`__.
```
Comment (ericl) [inline doc/source/autoscaling.rst]: simple-cluster.yaml and full-cluster.yaml?

File: python/ray/autoscaler/autoscaler.py [hunk 4]
```diff
@@ -18,77 +18,85 @@
...
 import ray.services as services
 
+REQUIRED, OPTIONAL = True, False
 
+# For (a, b), if a is a dictionary object, then
+# no extra fields can be introduced.
 CLUSTER_CONFIG_SCHEMA = {
     # An unique identifier for the head node and workers of this cluster.
-    "cluster_name": str,
+    "cluster_name": (str, REQUIRED),
```
Comment (ericl) [inline python/ray/autoscaler/autoscaler.py]: In the future, we could make this a protobuf, but it's not worth the complexity right now.

File: python/ray/autoscaler/autoscaler.py [hunk 5]
```diff
@@ -18,77 +18,85 @@
...
+    "provider": ({
+        "type": (str, REQUIRED),  # e.g. aws
+        "region": (str, REQUIRED),  # e.g. us-east-1
+        "availability_zone": (str, REQUIRED),  # e.g. us-east-1a
+    }, REQUIRED),
 
     # How Ray will authenticate with newly launched nodes.
-    "auth": dict,
+    "auth": ({
+        "ssh_user": (str, REQUIRED),  # e.g. aws
```
Comment (ericl) [inline python/ray/autoscaler/autoscaler.py]: e.g. ubuntu

File: python/ray/autoscaler/autoscaler.py [hunk 6]
```diff
@@ -18,77 +18,85 @@
...
+    "head_start_ray_commands": (list, OPTIONAL),
 
     # Command to start ray on worker nodes. You shouldn't need to modify this.
-    "worker_start_ray_commands": list,
+    "worker_start_ray_commands": (list, OPTIONAL),
 
     # Whether to avoid restarting the cluster during updates. This field is
     # controlled by the ray --no-restart flag and cannot be set by the user.
-    "no_restart": None,
+    "no_restart": (None, None),
```
Comment (ericl) [inline python/ray/autoscaler/autoscaler.py]: OPTIONAL?

File: python/ray/autoscaler/aws/minimal.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# An unique identifier for the head node and workers of this cluster.
+cluster_name: minimal
+
+# The maximum number of workers nodes to launch in addition to the head
+# node. This takes precedence over min_workers.
```
Comment (ericl) [inline python/ray/autoscaler/aws/minimal.yaml]: What's the min? Is it also max_workers or is it 0?

File: python/ray/tune/logger.py (hunk 1 above)
Comment (ericl) [reply python/ray/tune/logger.py]: Don't we have yaml as a Ray dep now, so it will be installed by default?

File: doc/source/autoscaling.rst (hunk 3 above)
Comment (richardliaw) [reply doc/source/autoscaling.rst]: So right now `default.yaml` is separate from `example.yaml`. Should they be the same?

Comment (ericl) [reply doc/source/autoscaling.rst]: How about we create `example-full.yaml` and `example-simple.yaml`, and symlink `example-full.yaml` to `cluster-defaults.yaml`

File: python/ray/autoscaler/aws/example.yaml [hunk 8]
```diff
@@ -84,7 +84,9 @@ setup_commands:
     # Note: if you're developing Ray, you probably want to create an AMI that
     # has your Ray repo pre-cloned. Then, you can replace the pip installs
     # below with a git checkout <your_sha> (and possibly a recompile).
-    - most_recent() { echo pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/$(aws s3 ls s3://ray-wheels --recursive | grep $1 | sort -r | head -n 1 | awk '{print $4}'); } && $( most_recent "cp36-cp36m-manylinux1" ) || $( most_recent "cp35-cp35m-manylinux1" )
+    # - pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.3.1-cp27-cp27mu-manylinux1_x86_64.whl
+    # - pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.3.1-cp35-cp35m-manylinux1_x86_64.whl
+    - pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.3.1-cp36-cp36m-manylinux1_x86_64.whl
```
Comment (ericl) [inline python/ray/autoscaler/aws/example.yaml]: Yeah please combine example.yaml and default.yaml, it's quite confusing.

**Review by ericl — APPROVED** [review]
This looks great! lgtm

**Comment by richardliaw** [comment]
test failures unrelated + one test is hanging; python2.7 and python3.6 have both passed, merging.
